### PR TITLE
feat(player): add debug logs

### LINKED ISSUE #702: MSP debugging events
Hi,

I'm wondering if it would be possible to include more debugging events?

Also, would it be possible to add the following kind of information to the client stream data overlay:

* Stream uptime
* Buffer delay
* Dropped frames

image

### PR DESCRIPTION
**Pull request by danielkaxis**
### Describe your changes

Show debug logs for delay, current time and buffered end. In our application we are experiencing
a big delay and these logs will be of help to debug this issue. 

### Issue ticket number and link

- Fixes #702 

### Checklist before requesting a review

- [x] I have performed a self-review of my own code
- [x] I have verified that the code builds perfectly fine on my local system
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have verified that my code follows the style already available in the repository
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by lekoaf — COMMENTED** [review]
Why reinvent the wheel?

https://github.com/debug-js/debug

File: player/src/HttpMp4Video.tsx [hunk 1]
```diff
@@ -95,6 +95,67 @@ export const HttpMp4Video: React.FC<HttpMp4VideoProps> = ({
...
+  /**
+   * Show debug logs with information received from
+   * 'progress' & 'timeupdate' events including the current
+   * up time, delay and end time of last buffer.
+   */
+  useEffect(() => {
+    const debugKey = localStorage.getItem('debug')
+    if (
+      // Don't initiate event listeners if debug is not on
+      debugKey !== 'msp:http-mp4-video' ||
```
Comment (lekoaf) [inline player/src/HttpMp4Video.tsx:107]: `debug` also supports `msp:*` for example, but that's not reflected here.

File: player/src/HttpMp4Video.tsx [hunk 2]
```diff
@@ -95,6 +95,67 @@ export const HttpMp4Video: React.FC<HttpMp4VideoProps> = ({
...
+    const onprogress = (event: Event) => {
+      const videoEl = event.currentTarget as HTMLVideoElement | null
+      if (videoEl === null) {
+        return
+      }
+
+      try {
+        const bufferedEnd = videoEl.buffered.end(videoEl.buffered.length - 1)
+        debugLog('%o', {
+          bufferedEnd,
```
Comment (lekoaf) [inline player/src/HttpMp4Video.tsx:123]: How is someone supposed to use this debug information?

I'd like to see something more concrete. Either in the debug log, as a comment here or in the documentation about how to interpret all the numbers and stuff that will come flying when they turn on debug.

**Review by lekoaf — COMMENTED** [review]
Add a section in the readme that this can be turned on and how.

File: player/src/HttpMp4Video.tsx (hunk 1 above)
Comment (danielkaxis) [reply player/src/HttpMp4Video.tsx]: Done

File: player/src/HttpMp4Video.tsx (hunk 2 above)
Comment (danielkaxis) [reply player/src/HttpMp4Video.tsx:129]: Done

**Review by lekoaf — APPROVED** [review]

File: player/src/HttpMp4Video.tsx [hunk 3]
```diff
@@ -95,6 +95,73 @@ export const HttpMp4Video: React.FC<HttpMp4VideoProps> = ({
...
+
+    if (
+      // Don't initiate event listeners if debug is not on
+      !shouldDebug ||
+      videoRef.current === null ||
+      !playing
+    ) {
+      return
+    }
+
```
Comment (steabert) [inline player/src/HttpMp4Video.tsx:119]: const videoEl = videoRef.current
if (videoEl === null) {
  return
}

File: player/src/HttpMp4Video.tsx [hunk 4]
```diff
@@ -95,6 +95,73 @@ export const HttpMp4Video: React.FC<HttpMp4VideoProps> = ({
...
+    const onprogress = (event: Event) => {
+      const videoEl = event.currentTarget as HTMLVideoElement | null
+      if (videoEl === null) {
+        return
+      }
```
Comment (steabert) [inline player/src/HttpMp4Video.tsx:124]: Not needed

File: player/src/HttpMp4Video.tsx [hunk 5]
```diff
@@ -95,6 +95,73 @@ export const HttpMp4Video: React.FC<HttpMp4VideoProps> = ({
...
+      } catch (err) {
+        debugLog('%o', err)
+      }
+    }
+
+    const ontimeupdate = (event: Event) => {
+      const videoEl = event.currentTarget as HTMLVideoElement | null
+      if (videoEl === null) {
+        return
+      }
```
Comment (steabert) [inline player/src/HttpMp4Video.tsx:140]: Don't need event or extra logic here.

File: player/src/WsRtspVideo.tsx [hunk 6]
```diff
@@ -137,6 +137,72 @@ export const WsRtspVideo: React.FC<WsRtspVideoProps> = ({
...
+
+    const videoEl = videoRef.current
+    videoEl.addEventListener('timeupdate', ontimeupdate)
+    videoEl.addEventListener('progress', onprogress)
+
+    return () => {
+      videoEl.removeEventListener('timeupdate', ontimeupdate)
+      videoEl.removeEventListener('progress', onprogress)
+    }
+  }, [playing])
```
Comment (steabert) [inline player/src/WsRtspVideo.tsx:204]: Same comments.

File: player/src/HttpMp4Video.tsx [hunk 7]
```diff
@@ -95,6 +95,73 @@ export const HttpMp4Video: React.FC<HttpMp4VideoProps> = ({
 
   const __sensorTmRef = useRef<TransformationMatrix>()
 
+  /**
```
Comment (steabert) [inline player/src/HttpMp4Video.tsx:98]: Extract as common hook.

File: player/src/HttpMp4Video.tsx [hunk 8]
```diff
@@ -62,7 +62,7 @@ export const HttpMp4Video: React.FC<HttpMp4VideoProps> = ({
   onEnded,
   metadataHandler,
 }) => {
-  let videoRef = useRef<HTMLVideoElement>(null)
+  let videoRef = useRef<HTMLVideoElement | null>(null)
```
[the remark this answers is not in the document]
Comment (danielkaxis) [reply player/src/HttpMp4Video.tsx]: Yup, not needed.

File: player/src/HttpMp4Video.tsx [hunk 9]
```diff
@@ -95,6 +95,73 @@ export const HttpMp4Video: React.FC<HttpMp4VideoProps> = ({
...
+      debugKey === 'msp:http-mp4-video' || debugKey === 'msp:*'
+
+    if (
+      // Don't initiate event listeners if debug is not on
+      !shouldDebug ||
+      videoRef.current === null ||
+      !playing
+    ) {
+      return
+    }
```
[the remark this answers is not in the document]
Comment (danielkaxis) [reply player/src/HttpMp4Video.tsx]: Done

File: player/src/HttpMp4Video.tsx (hunk 3 above)
Comment (danielkaxis) [reply player/src/HttpMp4Video.tsx]: Done

File: player/src/HttpMp4Video.tsx (hunk 4 above)
Comment (danielkaxis) [reply player/src/HttpMp4Video.tsx]: Done

File: player/src/HttpMp4Video.tsx (hunk 5 above)
Comment (danielkaxis) [reply player/src/HttpMp4Video.tsx]: Done

File: player/src/WsRtspVideo.tsx (hunk 6 above)
Comment (danielkaxis) [reply player/src/WsRtspVideo.tsx]: Done

File: player/src/HttpMp4Video.tsx (hunk 7 above)
Comment (danielkaxis) [reply player/src/HttpMp4Video.tsx]: Done

**Comment by steabert** [comment]
Is there enough info in the README already?

**Review by steabert — CHANGES_REQUESTED** [review]
LGTM, but it would be great if there is a small mention of this in the README.md.

You can use the streams/README.md as example, it has a brief debug explanation. It would also be good to mention detailed debugging strings as well in a table.

**Comment by danielkaxis** [comment]
> LGTM, but it would be great if there is a small mention of this in the README.md.
> 
> You can use the streams/README.md as example, it has a brief debug explanation. It would also be good to mention detailed debugging strings as well in a table.

Added info in README

File: player/README.md
Comment (steabert) [inline player/README.md:1]: ❤️

**Review by steabert — APPROVED** [review]

**Pull request merged**